### DISCUSSION TITLE
Restricting Entity Labels across Models

### DISCUSSION THREAD

**Discussion by source19069** [explosion/spaCy#8401] [Help: Coding & Implementations]
We are basing our training and training Docs on different models to determine which has the best performance. For example, let's say en_core_web_sm. Is there a way to use this model without using any of its named entities (PERSON, ORG etc)? All we would like to see in the trained model is the entity label that we've trained it for.

I have attached a copy of our training config file. In particular, I'd like to point out two changes we've made to achieve restricting entities (which of course, have not worked out):

[components.ner]
source = "en_core_web_sm" (This line was added, those below were commented.)
#factory = "ner"
#moves = null
#update_with_oracle_cut_size = 100

[initialize]
vectors = "en_core_web_sm"  (Only this line was changed)
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

Any guidance you can provide would be appreciated. Thank you!

[en_core_web_sm.config.txt](https://github.com/explosion/spaCy/files/6657921/en_core_web_sm.config.txt)

## Your Environment

 python -m spacy info --markdown

## Info about spaCy

- **spaCy version:** 3.0.1
- **Platform:** Windows-2012ServerR2-6.3.9600-SP0
- **Python version:** 3.7.6
- **Pipelines:** en_core_sci_lg (0.4.0), en_core_sci_md (0.4.0), en_core_sci_sm (0.4.0), en_core_web_lg (3.0.0), en_core_web_md (3.0.0), en_core_web_sm (3.0.0), en_core_web_trf (3.0.0), en_ner_bionlp13cg_md (0.4.0), en_pdl1_value (0.0.1), en_pdl1_value_trf (0.0.1)

**Comment by polm** [comment]
>  Is there a way to use this model without using any of its named entities (PERSON, ORG etc)?

What do you mean by "use" it? I guess you're training NER models and want to try different base models. In that case you should just put the `components.ner` section back the way it was. The way you have it now, the NER component is just loaded from the pretrained model. 

The model is distributed as one thing, but it's made up of different layers - tok2vec is the base that converts tokens to vectors, and then the tagger, parser, and NER components are separate pieces you can use or not use. Your `components.ner` section is explicitly using the NER from the pretrained model.

It can often make sense to continue training the pretrained NER instead of starting from scratch, but that's only really true if you are re-using the existing entities. If you aren't interested them in at all then you might as well just train the NER from scratch. 

To use the pretrained models for features while not re-using their NER layers, the most important thing is to specify them as the `init_tok2vec` param in the `initialize` part of the config. Specifying them for vectors may also be helpful, but only for the medium and large models - small and transformer models have no vectors.

**Reply by ericlinML** [reply]
Hi @polm ! I also work with @source19069 . We took your points about the init_tok2vec layer (thank you for the advice!) and tried to run with it. We've tried making that fix and started trying out other models. However, our results seemed essentially the same (listed below), so we were wondering if we had been writing our config file correctly. It's been quite confusing to read through the documentation about the config file (for a long time, we had mixed up the factory NER component with the pretrained en_core_web_sm NER component). 

Token level BILUO F1 scores for our new NER entity (from spacy's builtin evaluate CLI script):
Scores are reported from 5x5 nested randomized CV (25 models run against 5 test splits) ; scores are reported as mean (standard dev); [minimum, maximum]
1.  en_core_web_sm: 0.745 (0.034); [0.681, 0.79] 
2.  en_core_web_md: 0.744 (0.033);  [0.677, 0.797] 
3.  en_core_sci_sm:  0.757 (0.037); [0.647, 0.821] 
4.  en_core_sci_md: 0.743 (0.017); [0.683, 0.807] 
5.  en_core_sci_lg: 0.740 (0.028); [0.677, 0.785] 

I realize that the sciSpacy models are beyond your purview, but the performance seems so similar that I'm suspicious if we're incorrectly initializing the tok2vec layer. In fact, when we had varied our config files to not initialize the tok2vec (or so I had believed); we got the literal exact same score distribution (we were careful in setting the same seed).  For example, with our previous run (with the config file from source19069's post), we also got 0.745 (0.034); [0.681, 0.790]. Could you help us look this over and point out if we're doing something wrong?

(to a prior point, we're happy with updating the pretrained NER) 

[config_initial_train.txt](https://github.com/explosion/spaCy/files/6712007/config_initial_train.txt)

**Reply by polm** [reply]
So unfortunately it does look like this is not working the way you expect it to. 

You have specified your tok2vec init correctly, but you're sourcing the NER model from an existing pipeline (sm in the attached file). The pretrained pipeline NER models use a bundled tok2vec ("independent" as described [in the docs](https://spacy.io/usage/embeddings-transformers#embedding-layers-config)), so when you change your tok2vec init setting nothing happens. 

You can also tell this is happening if you look at your `[nlp]` section - your `pipeline` just has an NER component, so the tok2vec component you specify isn't even included. 

Also, zooming out a little, it doesn't make sense to continue training a pretrained NER while changing the tok2vec layer. When two layers are trained together they sort of gently fuse together, as their embeddings depend on each other. If you want to train a layer from scratch on top of a pretrained layer - like using the tok2vecs from pretrained pipelines as a source of features - that's fine, but taking the tok2vec layer away from a pretrained NER or other downstream layer and putting it with a new NER is like ripping flowers off a rosebush and putting them on another bush. It's not guaranteed to fail but it's doing things the hard way.

You should do one of these things:

1. Continue training a pretrained NER model. (Your config is set up for this now)
2. Train an NER model from scratch without pretrained tok2vec. (Use the quickstart for this.)
3. Train an NER model from scratch with pretrained tok2vec. (Use quickstart as a base and modify it by adding `init_tok2vec`.)

You haven't explained what kind of data you're training on except it seems that you're not interested in the entities in the pretrained models, so I would use steps 2 and 3.

**Reply by ericlinML** [reply]
We're looking at medical health records notes and looking to extract a specific antigen's blood lab value (called PD-L1) that would help oncologists choose which cancer treatment to choose. What the values we're trying to extract look like are something along the lines of ("100PERCENT", "positive", "low", "+" etc). You're correct that we're not interested in any other entities except if they somehow helped us extract this entity from a clinical note document. 

A small point to make back to your response - I think we did specify the same models across the config (at least we intended to keep all layers of our model to be the same as the pretrained model). 

Something I want to point out about the quickstart config file and the "python -m spacy init fill-config" command is that it seems to have problems with the preset choice of spacy.MultiHashEmbed.v2 which comes up under [components.tok2vec.model.embed] in the config file. I've produced this error on two separate Windows machines and have had two colleagues (who both use one of those machines) come up with the same error. 

When I run that command (ie. python -m spacy init fill-config baseConfig.cfg testConfig.cfg), it throws this error stating that MultiHashEmbed.v2 is not in the function registry. This doesn't make sense to me since it seems like it should have come installed with spaCy.  Are you aware of any fixes with this? We're debating just using HashEmbedCNN.v1 but am not sure if this would work the same way with the current vectors. Error is pasted below:

catalogue.RegistryError: [E893] Could not find function 'spacy.MultiHashEmbed.v2' in function registry 'architectures'. If you're using a custom function, make sure the code is available. If the function is provided by a third-party package, e.g. spacy-transformers, make sure the package is installed in your environment.

Available names: spacy-legacy.MaxoutWindowEncoder.v1, spacy-legacy.MishWindowEncoder.v1, spacy-legacy.TextCatEnsemble.v1, spacy-legacy.Tok2Vec.v1, spacy-transformers.Tok2VecTransformer.v1, spacy-transformers.TransformerListener.v1, spacy-transformers.TransformerModel.v1, spacy.CharacterEmbed.v1, spacy.EntityLinker.v1, spacy.HashEmbedCNN.v1, spacy.MaxoutWindowEncoder.v2, spacy.MishWindowEncoder.v2, spacy.MultiHashEmbed.v1, spacy.PretrainCharacters.v1, spacy.PretrainVectors.v1, spacy.Tagger.v1, spacy.TextCatBOW.v1, spacy.TextCatCNN.v1, spacy.TextCatEnsemble.v2, spacy.TextCatLowData.v1, spacy.Tok2Vec.v2, spacy.Tok2VecListener.v1, spacy.TorchBiLSTMEncoder.v1, spacy.TransitionBasedParser.v1, spacy.TransitionBasedParser.v2" 


Thank you for continuing to work with us!

**Reply by adrianeboyd** [reply]
If you upgrade to the most recent version of spacy v3.0, then `init fill-config` should work with the website quickstart.

The quickstart on the website is updated to use the most recent version of spacy in the background (to take advantage of bug fixes and other updates), so if you want to use it, then you'd need to update spacy in your local venv.

If you generate the initial config with `spacy init config` using your current version of spacy, then `spacy init fill-config` and other commands should work fine with the defaults provided for that exact spacy version.

The only difference between `spacy.MultiHashEmbed.v1` and `spacy.MultiHashEmbed.v2` is how tokens without static vectors are handled for the `tok2vec`. I don't think it should affect the pretraining steps, but we should probably take another look to make sure the pretraining architectures aren't affected by the same kind of bug.
